cpp/src/structs/utilities.cpp

```diff
@@ -93,11 +93,16 @@ struct flattened_table {
                              order col_order,
                              null_order col_null_order)
   {
-    if (nullability == column_nullability::FORCE || col.nullable()) {
-      // nullable columns could be required for comparisions such as joins
+    // Even if it is not required to extract the bitmask to a separate column,
+    // we should always do that if the structs column has any null element.
+    // In addition, we should check for null by calling to `has_nulls()`, not `nullable()`.
+    // This is because when comparing structs columns, if one column has bitmask while the other
+    // does not (and both columns do not have any null element) then flattening them using
+    // `nullable()` will result in tables with different number of columns.
```

Comment:
This explanation doesn't make sense to me. Don't you have the same problem if one column has `has_nulls() == true` and the other has `has_nulls() == false`?

The solution should be to use `column_nullability::FORCE` on both the LHS and RHS like @hyperbolic2346 did https://github.com/rapidsai/cudf/pull/7963

Reply:
When one column has null while the other doesn't, during calling to `search()`, `column_nullability::FORCE` will be passed in.

Reply:
Keep in mind that `has_nulls()` isn't always free, so the performance trade off isn't totally clear cut.